test/unit/test_config.py

```diff
@@ -27,8 +27,22 @@ def test_requires_config(self):
   def test_create(self):
     self.assertFalse(os.path.isfile(TMP_CONFIG_FILE))
     try:
-      self.config.create('foo')
+      self.config.create('bar')
       self.assertTrue(os.path.isfile(TMP_CONFIG_FILE))
-      self.assertEqual(self.config.project_name, 'foo')
+      self.assertEqual(self.config.project_name, 'bar')
     finally:
       os.remove(TMP_CONFIG_FILE)
+
+class TestConfig(HokusaiUnitTestCase):
+  def test_config_yaml_parsing(self):
+    self.assertEqual(config.config.project_name, 'foo')
+    self.assertEqual(config.config.pre_deploy, 'migrate.sh')
+    self.assertEqual(config.config.post_deploy, 'sh -c report.sh')
+
+  def test_environment_fallback(self):
```

Comment:
I think this is particularly important b/c, in the current implementation, the ENV variables never _override_ config values, but they can be used if the config values are nil.

I would argue that this is a bit surprising, and as such, warrants a test case.

Reply:
Yea I was thinking about how this may be counter-intuitive and may get us into problems, also considering the case where they are fallbacks for explicit command options...  Take the `git_remote` option, which is currently invoked when calling `hoksuai [staging|production] deploy {tag}` and while set to `--git-remote origin` on CI it's going to depend for a local user how they have configured their git remote.  I have artsy remotes as `upstream` yet others have it configured as `artsy`.  So in this case, I think it's better that environment variables take precedence over project-local config.

So perhaps a better order of precedence would be:
- Explicit command option flag (if invoked with the flag)
- Environment vars (if set)
- Project config (if set then basically a require)
- Defaults

Also @joeyAghion wdyt?

Reply:
I agree with this change in precedence, it's definitely more intuitive to me.

If we implement this change, we should probably rename `fallback_to_env` to `allow_env`, _or_ potentially removing that flag completely. I wonder if there's much harm in allowing someone to meaningfully set a `HOKUSAI_PRE_DEPLOY` ENV key. It would definitely be weird, but IDK if it's worth the code to explicitly disallow it.

Reply:
What do you think of sniffing out the remote name from the local git config upon invocation?

Here's what `git remote --verbose` returns for me locally within the hokusai project:
```
origin  git@github.com:dblandin/hokusai.git (fetch)
origin  git@github.com:dblandin/hokusai.git (push)
upstream        git@github.com:artsy/hokusai (fetch)
upstream        git@github.com:artsy/hokusai (push)
```

We could determine which remote targets an artsy repository by inspecting the targets.

Reply:
Actually we think we use either a configured remote like `origin`/`upstream` or a URI like `git@github.com:artsy/hokusai.git` when invoking these git commands https://github.com/artsy/hokusai/blob/master/hokusai/services/deployment.py#L46 so it shouldn't even have to resolve remote URIs to local aliases...

Reply:
Given this I now think we shouldn't check `git_remote` from env (as it may cause unintended side-effects from one project to another) and only accept it as an option or from the config file.

Reply:
Tested this and everything works as expected using a git URI rather than a remote alias.